### PR TITLE
[SMTChecker] Reset reference variables on assignment to a variable of reference type

### LINKED ISSUE #10847: [SMTChecker] False negative with assignment to state variables of reference type
In the following example:
```
pragma experimental SMTChecker;
contract A {
	int[] a;
	function f() public {
		require(a.length == 1 && a[0] == 1);
		int[] storage u = a;
		assert(u[0] == 1); // should hold
		int[] memory b = new int[](2);
		a = b;
		assert(u[0] == 1); // should fail
		A.a = b;
		assert(u[0] == 1); // should fail
	}

	function push_v(int x) public {
		a.push(x);
	}
}
```
SMTChecker currently reports assertion violation for the third assertion, but not the second one. The second assertion should also be reported as violated.

The problem is slightly different handling of assignments based on whether or not the contract type is used.

### PR DESCRIPTION
**Pull request by blishko**
Resolves #10847.

This is the second PR regarding handling of assignment in SMTChecker.
Previously, on an assignment to a reference variable, the knowledge about variables potentially aliasing the variable being assigned was not reset. This could lead to false negative as witnessed by #10847.

Unfortunately, this change introduces more false positives, as witneseds by the changed test. However, false positives are more acceptable than false negatives.

I plan to address the introduced false positive with the next PR in the series.

### CODE REVIEW

File: test/libsolidity/smtCheckerTests/types/bytes_2.sol [hunk 1]
```diff
@@ -7,3 +7,5 @@ contract C
 		assert(b1[1] == b2[1]);
 	}
 }
+// ----
+// Warning 6328: (119-141): CHC: Assertion violation happens here.\nCounterexample:\n\n\nTransaction trace:\nC.constructor()\nC.f(b1, b2)
```
Comment (blishko) [inline test/libsolidity/smtCheckerTests/types/bytes_2.sol:11]: @leonardoalt , basically, most of the code I have showed you today was to address this introduced false positive. I could add the code here, but I think it is better to do it gradually, so I would do it in another PR.

Comment (leonardoalt) [reply test/libsolidity/smtCheckerTests/types/bytes_2.sol:11]: Yep, makes sense!

File: test/libsolidity/smtCheckerTests/operators/assignment_contract_member_variable_array_3.sol [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+	int[] a;
+	function f() public {
+		require(a.length == 1 && a[0] == 1);
+		int[] storage u = a;
+		assert(u[0] == 1); // should hold
+		int[] memory b = new int[](2);
+		a = b;
+		assert(u[0] == 1); // should fail
+		A.a = b;
+		assert(u[0] == 1); // should fail
```
Comment (leonardoalt) [inline test/libsolidity/smtCheckerTests/operators/assignment_contract_member_variable_array_3.sol:12]: @blishko so this used to not fail before this PR?

Comment (blishko) [reply test/libsolidity/smtCheckerTests/operators/assignment_contract_member_variable_array_3.sol:12]: In the latest release one of them is detected and one is not.
After the fix #11088, actually not of these two failures is detected.

**Review by leonardoalt — APPROVED** [review]
